src/utils/resolve-ref.js

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const rp = require('request-promise-native');
+const URI = require('uri-js');
+
+const REGEXP_SHA = /^[0-9a-f]{40}$/i;
+
+/**
+ * Resolves the 'ref' of a github request if it's not already a SHA.
+ */
+async function resolveRef(context, { secrets = {}, request, logger }) {
+  if (!request || !request.params) {
+    throw new Error('Request parameters missing');
+  }
+  if (!secrets.REPO_RAW_ROOT) {
+    logger.debug('cannot resolve ref. No REPO_RAW_ROOT specified.');
+    return;
+  }
+  const rootURI = URI.parse(secrets.REPO_RAW_ROOT);
+  if (rootURI.host !== 'raw.github.com' && rootURI.host !== 'raw.githubusercontent.com') {
+    logger.warn(`unable to resolve ref to non-github repository: ${secrets.REPO_RAW_ROOT}`);
+    return;
+  }
+
+  const lookupUrl = secrets.API_RESOLVE_GIT_REF;
+  if (!lookupUrl) {
+    logger.debug('ignoring github ref resolving. API_RESOLVE_GIT_REF is not set.');
+    return;
+  }
+
+  // get required request parameters
+  const {
+    owner, repo,
+  } = request.params;
+
+  let { ref } = request.params;
+  if (!owner) {
+    throw new Error('Unknown owner, cannot resolve github ref');
+  }
+  if (!repo) {
+    throw new Error('Unknown repo, cannot resolve github ref');
+  }
+  if (!ref) {
+    logger.warn(`Recoverable error: no ref given for ${repo}/${owner}, falling back to master`);
+    ref = 'master';
+  }
+
+  if (REGEXP_SHA.test(ref)) {
+    logger.debug(`github-ref '${ref}' looks like a SHA. Ignoring resolving.`);
+    return;
+  }
+
+  try {
+    const result = await rp({
```

Comment:
I'd use the `openwhisk.action.invoke` function instead of `rp`

Reply:
I'd have it rather independent of openwhisk...or do you think because of performance?

Reply:
Also, maybe we should keep the naming convention in sync with the embed service, and name it `RESOLVE_GITREF_SERVICE` ?

Reply:
I think the OpenWhisk performance is better (and OpenWhisk can apply tweaks like running both actions in the same container, etc). I can live with having both, but think that's too much flexibility too early.